File: python/day_7/part_2.py
def solve(input: str) -> int:
    lines = input.strip().split("\n")
    grid = [list(line) for line in lines]
    rows = len(grid)
    cols = len(grid[0]) if rows > 0 else 0

    start_col = -1
    for col in range(cols):
        if grid[0][col] == "S":
            start_col = col
            break

    memo = {}

    def count_paths(row, col):
        if row >= rows:
            return 1

        if col < 0 or col >= cols:
            return 0

        if (row, col) in memo:
            return memo[(row, col)]

        if grid[row][col] == "^":
            result = count_paths(row + 1, col - 1) + count_paths(row + 1, col + 1)
        else:
            result = count_paths(row + 1, col)

        memo[(row, col)] = result
        return result

    return count_paths(0, start_col)


if __name__ == "__main__":
    with open("inputs/day_7.txt", "r") as f:
        input = f.read()
        result = solve(input)
        print(result)

File: go/day_7/part_2.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func solve(input string) int {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	grid := [][]rune{}
	for _, line := range lines {
		grid = append(grid, []rune(line))
	}
	rows := len(grid)
	cols := 0
	if rows > 0 {
		cols = len(grid[0])
	}

	start_col := -1
	for col := 0; col < cols; col++ {
		if grid[0][col] == 'S' {
			start_col = col
			break
		}
	}

	memo := map[[2]int]int{}

	var count_paths func(row, col int) int
	count_paths = func(row, col int) int {
		if row >= rows {
			return 1
		}

		if col < 0 || col >= cols {
			return 0
		}

		key := [2]int{row, col}
		if val, ok := memo[key]; ok {
			return val
		}

		result := 0
		if grid[row][col] == '^' {
			result = count_paths(row+1, col-1) + count_paths(row+1, col+1)
		} else {
			result = count_paths(row+1, col)
		}

		memo[key] = result
		return result
	}

	return count_paths(0, start_col)
}

func main() {
	input, err := os.ReadFile("../inputs/day_7.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
